### PR TITLE
Only send shipping information when all required fields are available

### LINKED ISSUE #576: Don't require address for renewal of virtual subscriptions
This is a continuation of/solution to this issue: https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/issues/386

## Synopsis

Renewal payments are failing for subscriptions containing virtual products which use the PayPal Checkout plugin. On renewals containing virtual subscription products, the PayPal Checkout plugin appears to be always requiring a billing address when requesting payment from PayPal.

## Details

When a customer purchases a virtual subscription product using the using the PayPal Checkout shortcut button on the product or cart page, and `require billing address` is not checked, an address is not passed back to WooCommerce, leaving the subscription itself without any address associated with it.

When an automatic renewal is triggered, the payment request is sent to PayPal with the `REQBILLINGADDRESS` param. This causes the payment to fail, because there is no address available. 

In other words, it appears that renewal payments are not honoring the `require billling address` option being unchecked.

## Expectation

When an automatic renewal  payment request is made by the PayPal Checkout plugin,  it should not include the `REQBILLINGADDRESS` param unless that requirement option is checked in the settings.

## Actions to be taken

1. Verify if that param is being sent with all renewal payments, or if billing address is being otherwise required on all renewals, even if the `require billing address` setting is not checked.
2. If it is, then either conditionally remove that param from the payment request, or just remove it from renewals altogether unless the `require billing address` setting is checked.

## Context

This has arisen from a phone conversation with Raúl Uriarte, jr. from PayPal and @EktaPayPal . 

This would be a potential solution to this issue: https://a8c.slack.com/archives/C100MSW86/p1558465858001400

Why is this a good solution?
- It would prevent the need for the 'require billing address' setting to be checked for the PayPal Checkout plugin to be compatible with Subscriptions, as noted here: https://docs.woocommerce.com/document/paypal-express-checkout/#section-17
- It would be a superior merchant experience, because they wouldn't need to talk with PayPal reps to get their plugin to work with Subscriptions for virtual products.
- Fewer addresses being passed around when not needed is better for customer privacy and security.

### PR DESCRIPTION
**Pull request by v18**
# Description of the issue

This fixes #576. The issue is that virtual subscriptions were not able to be renewed.

## Steps to reproduce the problem:
1. Check out the master branch and apply commit https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/commit/38802619bb6ef2a31a998017366c22bf436bd659 to this plugin
2. Create a virtual subscription
3. View the product page, click the PayPal Checkout shortcut purchase button
4. This will skip the checkout, redirect you to PayPal to approve the purchase, and return you to the site. After completion, a subscription will have been created for that customer
5. Edit the subscription from the admin area, see that there are no addresses in the address fields.
6. 'process renewal' as detailed here: https://docs.woocommerce.com/document/subscriptions/add-or-modify-a-subscription/#section-15
7. See that the renewal fails

If you add a breakpoint on exceptions or [here](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/includes/class-wc-gateway-ppec-with-paypal-addons.php#L224), you'll see that PayPal responds with an error:

```
ACK:"Failure"
L_ERRORCODE0:"10727"
L_ERRORCODE1:"10728"
L_ERRORCODE2:"10729"
L_ERRORCODE3:"10730"
L_ERRORCODE4:"10004"
L_SHORTMESSAGE0:"Shipping Address1 Empty"
L_SHORTMESSAGE1:"Shipping Address City Empty"
L_SHORTMESSAGE2:"Shipping Address State Empty"
L_SHORTMESSAGE3:"Shipping Address Postal Code Empty"
L_SHORTMESSAGE4:"Invalid Data"
L_LONGMESSAGE0:"The field Shipping Address1 is required"
L_LONGMESSAGE1:"The field Shipping Address City is required"
L_LONGMESSAGE2:"The field Shipping Address State is required"
L_LONGMESSAGE3:"The field Shipping Address Postal Code is required"
L_LONGMESSAGE4:"This transaction cannot be processed."
L_SEVERITYCODE0:"Error"
L_SEVERITYCODE1:"Error"
L_SEVERITYCODE2:"Error"
L_SEVERITYCODE3:"Error"
L_SEVERITYCODE4:"Error"
```

## Steps to reproduce the solution:
After having done the above steps, check out this branch, `try/fix-576` and follow these steps:

8. Reactivate the subscription from the previous steps
9. 'process renewal' as detailed here: https://docs.woocommerce.com/document/subscriptions/add-or-modify-a-subscription/#section-15
10. See that the renewal goes through

# Description of the solution

Before adding the `SHIPTO` parameters to the PayPal API call ([#](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/91c3aa7f1f88d1b0213d9624b7fee29503b8161f/includes/class-wc-gateway-ppec-client.php#L1015)), make sure that all of the required fields are available. The required fields are listed here:

https://developer.paypal.com/docs/classic/api/merchant/DoReferenceTransaction-API-Operation-NVP/

We can also remove the requirement that all virtual subscriptions have "Billing Address required" set. This was done in the revert commit https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/commit/38802619bb6ef2a31a998017366c22bf436bd659

~We may still need to fix the initial purchase of virtual subscriptions by sending a `REQBILLINGADDRESS` or `NOSHIPPING`=1 - but I don't think these are needed.~ Decided it's not needed.

One thing that works out really well with this fix is that all previous virtual subscriptions purchased without an address will work when renewed 🎉

# Alternative solutions considered

Since the affected subscriptions are virtual, we can also check for a virtual subscriptions. However, I opted for checking for all the required address fields, since that's was the reason for PayPal's error.

If we wanted to check for virtual subscriptions, we could use something like this [here](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/includes/class-wc-gateway-ppec-with-paypal-addons.php#L187):

```
$items = $order->get_items();
$is_all_items_virtual = array_reduce(
	$items,
	function ( $result, $item) {
		$product_id = $item->get_product_id();
		$product_variation_id = $item->get_variation_id();
		$canonical_product_id = ( 0 === $product_id ? $product_variation_id : $product_id );
		$product = wc_get_product( $canonical_product_id );
		$is_virtual = $product->is_virtual();
		return $result && $is_virtual;
	}
);
```

# To Test

No changes were made to how we're collecting payment the first time a subscription is purchased, so it should still work. To test thoroughly, though:

- [x] purchase one subscription using the normal flow & one using the express flow. try with virtual and non-virtual.
- [x] purchase a normal product using the normal flow & using the express flow
- [x] renew a subscription purchased using the normal flow. try with virtual and non-virtual.
- [x] renew a subscription purchased using the express flow. try with virtual and non-virtual.

@NickGreen let me know if you are able to test things out. Maybe also make sense to leave testing until we're ready to release a new version to reduce the testing burden.

### CODE REVIEW

**Comment by NickGreen** [comment]
I've run some tests locally on my test site. Steps and results:

1. Downloaded and activated https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/tree/try/fix-576
2. Clicked the PayPal Checkout shortcut button from the product page of a virtual subscription product
3. Signed into PayPal, agreed to purchase
4. Returned to WC site, click `Sign Up Now` button
5. Get this error at top of page: `PayPal error (10412): An error (10412) occurred while processing your PayPal payment. Please contact the store owner for assistance.`
6. Here's the whole set of error log entries:

```
06-27-2019 @ 23:17:38 - WC_Gateway_PPEC_Client::_request: remote request to https://api-3t.sandbox.paypal.com/nvp with params: Array
(
    [LOGOIMG] => 
    [HDRIMG] => 
    [PAGESTYLE] => 
    [BRANDNAME] => subtest1
    [RETURNURL] => http://subtest1.local/checkout/?woo-paypal-return=true&create-billing-agreement=true
    [CANCELURL] => http://subtest1.local/cart/?woo-paypal-cancel=true
    [LANDINGPAGE] => Login
    [SOLUTIONTYPE] => Sole
    [PAYMENTREQUEST_0_PAYMENTACTION] => sale
    [PAYMENTREQUEST_0_INSURANCEAMT] => 0
    [PAYMENTREQUEST_0_HANDLINGAMT] => 0
    [PAYMENTREQUEST_0_CUSTOM] => 
    [PAYMENTREQUEST_0_INVNUM] => 
    [PAYMENTREQUEST_0_CURRENCYCODE] => USD
    [PAYMENTREQUEST_0_AMT] => 1
    [PAYMENTREQUEST_0_ITEMAMT] => 1
    [PAYMENTREQUEST_0_SHIPPINGAMT] => 0
    [PAYMENTREQUEST_0_TAXAMT] => 0
    [PAYMENTREQUEST_0_SHIPDISCAMT] => 0
    [NOSHIPPING] => 1
    [EMAIL] => just@testing.com
    [L_BILLINGTYPE0] => MerchantInitiatedBillingSingleAgreement
    [L_BILLINGAGREEMENTDESCRIPTION0] => Orders with subtest1
    [L_BILLINGAGREEMENTCUSTOM0] => 
    [PAYMENTREQUEST_0_SHIPTONAME] => ProspressTest Buyer
    [PAYMENTREQUEST_0_SHIPTOSTREET] => test
    [PAYMENTREQUEST_0_SHIPTOSTREET2] => 
    [PAYMENTREQUEST_0_SHIPTOCITY] => test
    [PAYMENTREQUEST_0_SHIPTOSTATE] => AK
    [PAYMENTREQUEST_0_SHIPTOZIP] => 34554
    [PAYMENTREQUEST_0_SHIPTOCOUNTRYCODE] => US
    [PAYMENTREQUEST_0_SHIPTOPHONENUM] => 3453455
    [L_PAYMENTREQUEST_0_NAME0] => virtual sub test
    [L_PAYMENTREQUEST_0_DESC0] => 
    [L_PAYMENTREQUEST_0_QTY0] => 1
    [L_PAYMENTREQUEST_0_AMT0] => 1
    [METHOD] => SetExpressCheckout
    [VERSION] => 120.0
    [USER] => [redacted]
    [PWD] => [redacted]
    [SIGNATURE] => [redacted]
)

06-27-2019 @ 23:17:40 - WC_Gateway_PPEC_Client::_process_response: acknowleged response body: Array
(
    [TOKEN] => EC-0XP37782MG628230B
    [TIMESTAMP] => 2019-06-27T23:17:40Z
    [CORRELATIONID] => 446274edad4b4
    [ACK] => Success
    [VERSION] => 120.0
    [BUILD] => 53072421
)

06-27-2019 @ 23:18:49 - WC_Gateway_PPEC_Client::_request: remote request to https://api-3t.sandbox.paypal.com/nvp with params: Array
(
    [METHOD] => GetExpressCheckoutDetails
    [VERSION] => 120.0
    [TOKEN] => EC-0XP37782MG628230B
    [USER] => [redacted]
    [PWD] => [redacted]
    [SIGNATURE] => [redacted]

06-27-2019 @ 23:18:50 - WC_Gateway_PPEC_Client::_process_response: acknowleged response body: Array
(
    [TOKEN] => EC-0XP37782MG628230B
    [BILLINGAGREEMENTACCEPTEDSTATUS] => 1
    [CHECKOUTSTATUS] => PaymentActionNotInitiated
    [TIMESTAMP] => 2019-06-27T23:18:49Z
    [CORRELATIONID] => e8fbd7d9f0c1
    [ACK] => Success
    [VERSION] => 120.0
    [BUILD] => 53072421
    [EMAIL] => [redacted]
    [PAYERID] => P6H7MH7A8VFA6
    [PAYERSTATUS] => verified
    [FIRSTNAME] => ProspressTest
    [LASTNAME] => Buyer
    [COUNTRYCODE] => US
    [SHIPTOPHONENUM] => 3453455
    [ADDRESSSTATUS] => Confirmed
    [CURRENCYCODE] => USD
    [AMT] => 1.00
    [ITEMAMT] => 1.00
    [SHIPPINGAMT] => 0.00
    [HANDLINGAMT] => 0.00
    [TAXAMT] => 0.00
    [NOTIFYURL] => [redacted]
    [INSURANCEAMT] => 0.00
    [SHIPDISCAMT] => 0.00
    [INSURANCEOPTIONOFFERED] => false
    [L_NAME0] => virtual sub test
    [L_QTY0] => 1
    [L_TAXAMT0] => 0.00
    [L_AMT0] => 1.00
    [PAYMENTREQUEST_0_CURRENCYCODE] => USD
    [PAYMENTREQUEST_0_AMT] => 1.00
    [PAYMENTREQUEST_0_ITEMAMT] => 1.00
    [PAYMENTREQUEST_0_SHIPPINGAMT] => 0.00
    [PAYMENTREQUEST_0_HANDLINGAMT] => 0.00
    [PAYMENTREQUEST_0_TAXAMT] => 0.00
    [PAYMENTREQUEST_0_NOTIFYURL] => [redacted]
    [PAYMENTREQUEST_0_INSURANCEAMT] => 0.00
    [PAYMENTREQUEST_0_SHIPDISCAMT] => 0.00
    [PAYMENTREQUEST_0_SELLERPAYPALACCOUNTID] => [redacted]
    [PAYMENTREQUEST_0_INSURANCEOPTIONOFFERED] => false
    [PAYMENTREQUEST_0_SHIPTOPHONENUM] => 3453455
    [L_PAYMENTREQUEST_0_NAME0] => virtual sub test
    [L_PAYMENTREQUEST_0_QTY0] => 1
    [L_PAYMENTREQUEST_0_TAXAMT0] => 0.00
    [L_PAYMENTREQUEST_0_AMT0] => 1.00
    [PAYMENTREQUESTINFO_0_ERRORCODE] => 0
)

06-27-2019 @ 23:18:50 - WC_Gateway_PPEC_Client::_request: remote request to https://api-3t.sandbox.paypal.com/nvp with params: Array
(
    [METHOD] => GetExpressCheckoutDetails
    [VERSION] => 120.0
    [TOKEN] => EC-0XP37782MG628230B
    [USER] => [redacted]
    [PWD] => [redacted]
    [SIGNATURE] => [redacted]
)

06-27-2019 @ 23:18:51 - WC_Gateway_PPEC_Client::_process_response: acknowleged response body: Array
(
    [TOKEN] => EC-0XP37782MG628230B
    [BILLINGAGREEMENTACCEPTEDSTATUS] => 1
    [CHECKOUTSTATUS] => PaymentActionNotInitiated
    [TIMESTAMP] => 2019-06-27T23:18:50Z
    [CORRELATIONID] => 3ca7f39ea0e6c
    [ACK] => Success
    [VERSION] => 120.0
    [BUILD] => 53072421
    [EMAIL] => [redacted]
    [PAYERID] => P6H7MH7A8VFA6
    [PAYERSTATUS] => verified
    [FIRSTNAME] => ProspressTest
    [LASTNAME] => Buyer
    [COUNTRYCODE] => US
    [SHIPTOPHONENUM] => 3453455
    [ADDRESSSTATUS] => Confirmed
    [CURRENCYCODE] => USD
    [AMT] => 1.00
    [ITEMAMT] => 1.00
    [SHIPPINGAMT] => 0.00
    [HANDLINGAMT] => 0.00
    [TAXAMT] => 0.00
    [NOTIFYURL] => [redacted]
    [INSURANCEAMT] => 0.00
    [SHIPDISCAMT] => 0.00
    [INSURANCEOPTIONOFFERED] => false
    [L_NAME0] => virtual sub test
    [L_QTY0] => 1
    [L_TAXAMT0] => 0.00
    [L_AMT0] => 1.00
    [PAYMENTREQUEST_0_CURRENCYCODE] => USD
    [PAYMENTREQUEST_0_AMT] => 1.00
    [PAYMENTREQUEST_0_ITEMAMT] => 1.00
    [PAYMENTREQUEST_0_SHIPPINGAMT] => 0.00
    [PAYMENTREQUEST_0_HANDLINGAMT] => 0.00
    [PAYMENTREQUEST_0_TAXAMT] => 0.00
    [PAYMENTREQUEST_0_NOTIFYURL] => [redacted]
    [PAYMENTREQUEST_0_INSURANCEAMT] => 0.00
    [PAYMENTREQUEST_0_SHIPDISCAMT] => 0.00
    [PAYMENTREQUEST_0_SELLERPAYPALACCOUNTID] => shared-reftrans@mailinator.com
    [PAYMENTREQUEST_0_INSURANCEOPTIONOFFERED] => false
    [PAYMENTREQUEST_0_SHIPTOPHONENUM] => 3453455
    [L_PAYMENTREQUEST_0_NAME0] => virtual sub test
    [L_PAYMENTREQUEST_0_QTY0] => 1
    [L_PAYMENTREQUEST_0_TAXAMT0] => 0.00
    [L_PAYMENTREQUEST_0_AMT0] => 1.00
    [PAYMENTREQUESTINFO_0_ERRORCODE] => 0
)

06-27-2019 @ 23:19:02 - WC_Gateway_PPEC_Client::_request: remote request to https://api-3t.sandbox.paypal.com/nvp with params: Array
(
    [METHOD] => GetExpressCheckoutDetails
    [VERSION] => 120.0
    [TOKEN] => EC-0XP37782MG628230B
    [USER] => [redacted]
    [PWD] => [redacted]
    [SIGNATURE] => [redacted]
)

06-27-2019 @ 23:19:03 - WC_Gateway_PPEC_Client::_process_response: acknowleged response body: Array
(
    [TOKEN] => EC-0XP37782MG628230B
    [BILLINGAGREEMENTACCEPTEDSTATUS] => 1
    [CHECKOUTSTATUS] => PaymentActionNotInitiated
    [TIMESTAMP] => 2019-06-27T23:19:03Z
    [CORRELATIONID] => 67f845661503f
    [ACK] => Success
    [VERSION] => 120.0
    [BUILD] => 53072421
    [EMAIL] => [redacted]
    [PAYERID] => P6H7MH7A8VFA6
    [PAYERSTATUS] => verified
    [FIRSTNAME] => ProspressTest
    [LASTNAME] => Buyer
    [COUNTRYCODE] => US
    [SHIPTOPHONENUM] => 3453455
    [ADDRESSSTATUS] => Confirmed
    [CURRENCYCODE] => USD
    [AMT] => 1.00
    [ITEMAMT] => 1.00
    [SHIPPINGAMT] => 0.00
    [HANDLINGAMT] => 0.00
    [TAXAMT] => 0.00
    [NOTIFYURL] => [redacted]
    [INSURANCEAMT] => 0.00
    [SHIPDISCAMT] => 0.00
    [INSURANCEOPTIONOFFERED] => false
    [L_NAME0] => virtual sub test
    [L_QTY0] => 1
    [L_TAXAMT0] => 0.00
    [L_AMT0] => 1.00
    [PAYMENTREQUEST_0_CURRENCYCODE] => USD
    [PAYMENTREQUEST_0_AMT] => 1.00
    [PAYMENTREQUEST_0_ITEMAMT] => 1.00
    [PAYMENTREQUEST_0_SHIPPINGAMT] => 0.00
    [PAYMENTREQUEST_0_HANDLINGAMT] => 0.00
    [PAYMENTREQUEST_0_TAXAMT] => 0.00
    [PAYMENTREQUEST_0_NOTIFYURL] => [redacted]
    [PAYMENTREQUEST_0_INSURANCEAMT] => 0.00
    [PAYMENTREQUEST_0_SHIPDISCAMT] => 0.00
    [PAYMENTREQUEST_0_SELLERPAYPALACCOUNTID] => [redacted]
    [PAYMENTREQUEST_0_INSURANCEOPTIONOFFERED] => false
    [PAYMENTREQUEST_0_SHIPTOPHONENUM] => 3453455
    [L_PAYMENTREQUEST_0_NAME0] => virtual sub test
    [L_PAYMENTREQUEST_0_QTY0] => 1
    [L_PAYMENTREQUEST_0_TAXAMT0] => 0.00
    [L_PAYMENTREQUEST_0_AMT0] => 1.00
    [PAYMENTREQUESTINFO_0_ERRORCODE] => 0
)

06-27-2019 @ 23:19:03 - WC_Gateway_PPEC_Client::_request: remote request to https://api-3t.sandbox.paypal.com/nvp with params: Array
(
    [METHOD] => CreateBillingAgreement
    [VERSION] => 120.0
    [TOKEN] => EC-0XP37782MG628230B
    [USER] => [redacted]
    [PWD] => [redacted]
    [SIGNATURE] => [redacted]
)

06-27-2019 @ 23:19:04 - WC_Gateway_PPEC_Client::_process_response: acknowleged response body: Array
(
    [BILLINGAGREEMENTID] => B-60702816EX054511N
    [TIMESTAMP] => 2019-06-27T23:19:04Z
    [CORRELATIONID] => b8198a9bd267e
    [ACK] => Success
    [VERSION] => 120.0
    [BUILD] => 53072421
)

06-27-2019 @ 23:19:04 - WC_Gateway_PPEC_Client::_request: remote request to https://api-3t.sandbox.paypal.com/nvp with params: Array
(
    [TOKEN] => EC-0XP37782MG628230B
    [PAYERID] => P6H7MH7A8VFA6
    [PAYMENTREQUEST_0_AMT] => 1
    [PAYMENTREQUEST_0_ITEMAMT] => 1
    [PAYMENTREQUEST_0_SHIPPINGAMT] => 0
    [PAYMENTREQUEST_0_TAXAMT] => 0
    [PAYMENTREQUEST_0_SHIPDISCAMT] => 0
    [PAYMENTREQUEST_0_INSURANCEAMT] => 0
    [PAYMENTREQUEST_0_HANDLINGAMT] => 0
    [PAYMENTREQUEST_0_CURRENCYCODE] => USD
    [PAYMENTREQUEST_0_NOTIFYURL] => http://subtest1.local/wc-api/WC_Gateway_PPEC/
    [PAYMENTREQUEST_0_PAYMENTACTION] => sale
    [PAYMENTREQUEST_0_INVNUM] => WC-512
    [PAYMENTREQUEST_0_CUSTOM] => {"order_id":512,"order_number":"512","order_key":"wc_order_5mrCuBdVg0jso"}
    [NOSHIPPING] => 1
    [L_PAYMENTREQUEST_0_NAME0] => virtual sub test
    [L_PAYMENTREQUEST_0_DESC0] => 
    [L_PAYMENTREQUEST_0_QTY0] => 1
    [L_PAYMENTREQUEST_0_AMT0] => 1
    [METHOD] => DoExpressCheckoutPayment
    [VERSION] => 120.0
    [BUTTONSOURCE] => WooThemes_EC
    [USER] => [redacted]
    [PWD] => [redacted]
    [SIGNATURE] => [redacted]

06-27-2019 @ 23:19:06 - WC_Gateway_PPEC_Client::_process_response: acknowleged response body: Array
(
    [TOKEN] => EC-0XP37782MG628230B
    [SUCCESSPAGEREDIRECTREQUESTED] => false
    [TIMESTAMP] => 2019-06-27T23:19:06Z
    [CORRELATIONID] => f66ce64abf990
    [ACK] => Failure
    [VERSION] => 120.0
    [BUILD] => 53072421
    [L_ERRORCODE0] => 10412
    [L_SHORTMESSAGE0] => Duplicate invoice
    [L_LONGMESSAGE0] => Payment has already been made for this InvoiceID.
    [L_SEVERITYCODE0] => Error
)
```

**Comment by v18** [comment]
Thanks @NickGreen !

> Get this error at top of page: PayPal error (10412): An error (10412) occurred while processing your PayPal payment. Please contact the store owner for assistance.

> This appears to have to do with duplicate invoice numbers that get sent to PayPal.

Is this a known issue?

I tried to check out with a virtual subscription using the same steps on my test site and it worked for me just now 🤔. I'll dig in more & reply back.

**Comment by jrodger** [comment]
Is this the invoice number do you think? `[PAYMENTREQUEST_0_INVNUM] => WC-512`. It looks like that is generated on our side. Is it possible that PayPal account has seen that invoice ID before from another test site?

I haven't tried this, but it looks like you can disable that behaviour here when logged into the seller's PayPal account: https://www.sandbox.paypal.com/businessmanage/preferences/payments

File: includes/class-wc-gateway-ppec-client.php [hunk 1]
```diff
@@ -1012,7 +1012,37 @@ public function get_do_reference_transaction_params( array $args ) {
 			) ),
 		);
 
-		if ( ! empty( $details['shipping_address'] ) ) {
+		if ( ! empty( $details['shipping_address'] ) &&
```
Comment (jrodger) [inline includes/class-wc-gateway-ppec-client.php]: I found this a bit hard to get my head around. Would creating a new method help, or would that confuse things more?

```
if ( ! empty( $details['shipping_address'] ) && $details['shipping_address']->is_complete_shipping_address() ) {
```

Maybe on the address object, or if it feels too specific to `get_do_reference_transaction_params` then it could just be a seperate function in the `WC_Gateway_PPEC_Client` class?

File: includes/class-wc-gateway-ppec-client.php [hunk 2]
```diff
@@ -1012,7 +1012,37 @@ public function get_do_reference_transaction_params( array $args ) {
...
+			! empty( $details['shipping_address']->getName() ) &&
+			! empty( $details['shipping_address']->getStreet1() ) &&
+			! empty( $details['shipping_address']->getCity() ) &&
+			! empty( $details['shipping_address']->getCountry() ) &&
+			// If the country is the US, a zipcode is required
```
Comment (jrodger) [inline includes/class-wc-gateway-ppec-client.php]: Where is this zip code and mandatory state validation logic coming from? Just want to check we don't miss anything else that could be mandatory.

**Comment by jrodger** [comment]
I've followed the testing steps you outlined and tested the scenarios in the "To Test" section. Everything works well :+1:. It's a tidy fix.

Just a couple of code review comments.

File: includes/class-wc-gateway-ppec-client.php (hunk 2 above)
Comment (v18) [reply includes/class-wc-gateway-ppec-client.php]: I got all the requirements from this doc:

https://developer.paypal.com/docs/classic/api/merchant/DoReferenceTransaction-API-Operation-NVP/#ship-to-address-fields

For example, for state it says:

> It is required for transactions only if the address is in one of the following countries: Argentina, Brazil, Canada, China, Indonesia, India, Japan, Mexico, Thailand or USA. See the list of PayPal State codes.

Comment (jrodger) [reply includes/class-wc-gateway-ppec-client.php]: Ah great, straight for PayPal. Works for me :)

**Comment by NickGreen** [comment]
@jrodger Thanks for pointing that out. I think you're right - we use this sandbox account on lots of things, so it's likely that invoice has already been created.

I turned off that setting and the order completed properly, without errors.

A subscription was made with no address information: https://d.pr/free/i/bI2wD5

A renewal was successfully completed without error: https://d.pr/free/i/Lh5SfD

LGTM! 👍

**Comment by jrodger** [comment]
> When trying to pay for a manual renewal or when paying for a manual renewal early (i.e. running through the checkout process to pay for a renewal), I'm getting an API error: An error occurred while calling the PayPal API.

I get the same thing, but I haven't been able to establish if this is definitely down to the changes in this PR.

The failing request / response seems to be this:

```
07-02-2019 @ 16:59:07 - WC_Gateway_PPEC_Client::_request: remote request to https://api-3t.sandbox.paypal.com/nvp with params: Array
(
    [METHOD] => GetExpressCheckoutDetails
    [VERSION] => 120.0
    [TOKEN] => EC-07H89975X7859604D
    [USER] => james.rodger-facilitator_api1.automattic.com
    [PWD] => 3XLWKFCBYN7J6W7N
    [SIGNATURE] => A58M-aKN4f9LRWMkkNGF5Q.n6nM4AgIh5sjZBbv7zKhR24STzaQhY1KG
)

07-02-2019 @ 16:59:08 - WC_Gateway_PPEC_Client::_process_response: acknowleged response body: Array
(
    [TOKEN] => EC-07H89975X7859604D
    [BILLINGAGREEMENTACCEPTEDSTATUS] => 0
    [CHECKOUTSTATUS] => PaymentActionNotInitiated
    [TIMESTAMP] => 2019-07-02T16:59:08Z
    [CORRELATIONID] => 3f53c4ddd4390
    [ACK] => Success
    [VERSION] => 120.0
    [BUILD] => 53072421
    ...
    [L_NAME0] => Shippable Subscription
    [L_QTY0] => 1
    [L_TAXAMT0] => 0.00
    [L_AMT0] => 1.00
    [PAYMENTREQUEST_0_CURRENCYCODE] => USD
    [PAYMENTREQUEST_0_AMT] => 1.00
    [PAYMENTREQUEST_0_ITEMAMT] => 1.00
    [PAYMENTREQUEST_0_SHIPPINGAMT] => 0.00
    [PAYMENTREQUEST_0_HANDLINGAMT] => 0.00
    [PAYMENTREQUEST_0_TAXAMT] => 0.00
    [PAYMENTREQUEST_0_INSURANCEAMT] => 0.00
    [PAYMENTREQUEST_0_SHIPDISCAMT] => 0.00
    ...
    [PAYMENTREQUESTINFO_0_ERRORCODE] => 0
)
```

I think the `PaymentActionNotInitiated` could be the issue here, but haven't taken a closer look just yet.

**Comment by v18** [comment]
@NickGreen that's awesome, thanks so much for finding that problem.

I was able to repro the problem on my end. I also checked it using a commit before these changes were introduced (used 054e0f3cd1e03e249da009d2f121c1b689f4f815 since we needed to go back to a time when PayPal was an allowed gateway for virtual subs) - and the problem still happened then. 
So, the changes in this PR are not the root cause of the issue.

---

Here's a bit of digging:

We say that we need to create a billing subscription here:

https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/cc2943054c3726e445781af7b3f43d84a0e12d53/includes/abstracts/abstract-wc-gateway-ppec.php#L140

But when trying to create it, we see that the billing agreement has not been accepted here:

https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/includes/class-wc-gateway-ppec-checkout-handler.php#L838

---

Should we hold up this PR and/or the release to fix this?

**Comment by v18** [comment]
One way to solve this is by requiring a billing agreement when we first contact PayPal. You can try this out by using `return true` here and going through the flow:

https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/includes/class-wc-gateway-ppec-checkout-handler.php#L1021

When we first contact PayPal when we start checkout ([here](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/includes/class-wc-gateway-ppec-checkout-handler.php#L699)), we check whether we need to create a new billing agreement [here ](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/includes/class-wc-gateway-ppec-checkout-handler.php#L1020).

Ah!

So... here's what's up:

To see if we need a billing agreement, we check to see if `cart_contains_subscription` here:

https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/includes/class-wc-gateway-ppec-checkout-handler.php#L1025

This function comes from WooCommerce Subscriptions, and has this logic:

```
	public static function cart_contains_subscription() {

		$contains_subscription = false;

		if ( ! empty( WC()->cart->cart_contents ) && ! wcs_cart_contains_renewal() ) {
			foreach ( WC()->cart->cart_contents as $cart_item ) {
				if ( WC_Subscriptions_Product::is_subscription( $cart_item['data'] ) ) {
					$contains_subscription = true;
					break;
				}
			}
		}

		return $contains_subscription;
	}
```

So, it checks to see if we have a renewal on our hands first - `wcs_cart_contains_renewal()`. And since this is indeed a renewal, it returns `false` for `cart_contains_subscription()`.

However, PPEC knows we need to create a billing agreement ID if it's a renewal. That's encoded here:

https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/includes/class-wc-gateway-ppec-checkout-handler.php#L1032

`wcs_order_contains_renewal( $args['order_id'] )`

Though we only check that if we have an `order_id`.

It seems like PPEC does expect that we'll need to create a billing agreement, but we don't know about it.

We could check for it when `order_id` is not available using `wcs_cart_contains_renewal()`.

The code to check for renewal is 3 years old and `cart_contains_subscription()` hasn't been changed in longer than that, so this seems like a super rare issue with users renewing manually using PPEC.

**tldr: when deciding whether or not we need to create a billing agreement, we likely _do_ want to create it when a subscription is being renewed, and we could check for that using `wcs_cart_contains_renewal`.**

**Comment by v18** [comment]
Upon further reflection, I think that fix makes sense and I'm going to add it in tomorrow, as well as the function change 👍 .

**Comment by v18** [comment]
Fixed the issue with renewing the subscription manually. Ready to review again.

File: includes/class-wc-gateway-ppec-client.php (hunk 1 above)
Comment (v18) [reply includes/class-wc-gateway-ppec-client.php]: Thanks for pointing it out. Wasn't sure off the bat whether it was unreadable.

Added it to the address class - turns out many different API calls have the same shipping requirements so I didn't have to call it `has_all_required_shipping_params_for_do_reference_transaction` 😄

**Comment by jrodger** [comment]
I've re-test manually renewing a subscription, that works well for me now.

Code changes look good, like how you've done the address check method. It reads well.

If @NickGreen is happy then we can 🚢

**Comment by NickGreen** [comment]
I encountered no errors when going through the above tests with the most recent version of the `try/fix-576` branch, using a sandbox PayPal account. 

Any idea when the next release is planned for? I can spread this info to the support team. Thanks for your work so far!

**Comment by allendav** [comment]
@NickGreen - looks like we will release PP Checkout next week - we want to also try to include a fix for this issue: https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/issues/561#issuecomment-508274372
